Refactor SecureStorage into an interface

antonfirsov opened the pull request:
## Description of the change

Contributes to #4497.

## Additions made

```C#
public interface ISecureStorage
{
	Task<string> GetAsync(string key);
	Task SetAsync(string key, string value);
	bool Remove(string key);
	void RemoveAll();
}

public class SecureStorageImpl : ISecureStorage{ /*... */ }
```

antonfirsov committed:
refactor SecureStorage to use interfaces

antonfirsov committed:
do not use private keyword

rmarinho commented:
```
         D:\a\_work\1\s\src\Essentials\test\DeviceTests\Tests\SecureStorage_Tests.cs(32,33): error CS1501: No overload for method 'SetAsync' takes 3 arguments [D:\a\_work\1\s\src\Essentials\test\DeviceTests\Essentials.DeviceTests.csproj]
```

Build failing

antonfirsov committed:
Move iOS-specific API-s back to SecureStorage

antonfirsov committed:
move SecureStorageImplementation to Microsoft.Maui.Essentials.Implementations

antonfirsov commented on src/Essentials/src/SecureStorage/SecureStorage.ios.tvos.watchos.macos.cs:
```diff
@@ -9,8 +9,24 @@ namespace Microsoft.Maui.Essentials
+		public static Task SetAsync(string key, string value, SecAccessible accessible) =>
+			Implementations.SecureStorageImplementation.SetAsync(key, value, accessible);
+	}
+}
+
+namespace Microsoft.Maui.Essentials.Implementations
```
Here and in #4630 I had to make such namespace splits, since the static class doesn't live in the `Implementations` namespace. It seemed unworthy to split up the files, but if this is an unprecedented code pattern maybe I should do it?

@Redth Thoughts?

Redth replied:
I think it's fine in this case to have both in the same file :)

antonfirsov commented:
@rmarinho do we need clean CI on this and other Essentials PR-s? The failures I see so far seem to be unrelated.

rmarinho commented:
![image](https://user-images.githubusercontent.com/1235097/154142501-3d2d9a71-d6b5-49eb-8b7e-1a119028e975.png)

seems some tests are failing , do they work locally ?

antonfirsov committed:
prevent SecureStorage to throw during type initialization in netstandard

antonfirsov commented:
@rmarinho thanks, addressed! How to navigate to the view you pasted? I usually go here:

![image](https://user-images.githubusercontent.com/6835152/154147206-2f6b97a3-4a70-41fc-b0cc-b96a1ed17eb3.png)

but none of the checks I've seen listed test failures.

rmarinho commented:
You can go to the Tests tab

![image](https://user-images.githubusercontent.com/1235097/154161680-f5ddd28b-6a3a-4ecc-9492-4337b3ddc2ac.png)

mattleibow commented on src/Essentials/src/SecureStorage/SecureStorage.shared.cs:
```diff
@@ -27,15 +40,27 @@ public static Task SetAsync(string key, string value)
 			if (value == null)
 				throw new ArgumentNullException(nameof(value));
```
Can we move this logic into each platform? This way the static API is very very thin.

antonfirsov replied:
This is better be done after reaching API stability.

mattleibow commented on src/Essentials/src/SecureStorage/SecureStorage.ios.tvos.watchos.macos.cs:
```diff
@@ -9,8 +9,24 @@ namespace Microsoft.Maui.Essentials
+			set { Implementations.SecureStorageImplementation.DefaultAccessible = value; }
+		}
+
+		public static Task SetAsync(string key, string value, SecAccessible accessible) =>
+			Implementations.SecureStorageImplementation.SetAsync(key, value, accessible);
```
Maybe put this on `ISecAccessibleSecureStorage : ISecureStorage`

mattleibow replied:
Maybe if `Current is not ISecAccessibleSecureStorage` then we can throw a `NotImplementedException`.

antonfirsov replied:
Addressed in 0c43ef1. Note that `ISecAccessibleSecureStorage` has to be a platform-dependent type, otherwise we would have to mirror the `SecAccessible` enum as a "cross-platform" type.

mattleibow commented on src/Essentials/src/SecureStorage/SecureStorage.ios.tvos.watchos.macos.cs:
```diff
@@ -9,8 +9,24 @@ namespace Microsoft.Maui.Essentials
 	{
-		public static SecAccessible DefaultAccessible { get; set; } =
-			SecAccessible.AfterFirstUnlock;
+		public static SecAccessible DefaultAccessible
+		{
+			get { return Implementations.SecureStorageImplementation.DefaultAccessible; }
+			set { Implementations.SecureStorageImplementation.DefaultAccessible = value; }
+		}
```
And this too maybe. Feels a bit weird, but better to have this all on `(Current as ISecAccessibleSecureStorage ).DefaultAccessible`

antonfirsov committed:
Merge branch 'main' into refactor/SecureStorage

antonfirsov committed:
ISecAccessibleSecureStorage

Redth committed:
Merge branch 'main' into pr/4668

Redth committed:
Refactor to match other platform API's

In a few other PR's for Essentials refactoring we started adding the `IPlatformX` interface as a place to host platform specific things.

This adjusts to the same pattern used there.

Redth committed:
Expose alias to internal again
